File: src/tools/grep.ts
import { z } from 'zod';
import { execa } from 'execa';
import type { Tool } from './types';

const GrepParams = z.object({
  pattern: z.string().min(1).describe('Regular expression pattern to search for'),
  path: z.string().optional().describe('Directory or file to search in. Defaults to current working directory.'),
  glob: z.string().optional().describe('Glob pattern to filter files (e.g. "**/*.ts", "*.tsx")'),
  output_mode: z.enum(['content', 'files_with_matches', 'count']).optional().default('content'),
  '-i': z.boolean().optional().describe('Case insensitive search'),
  '-n': z.boolean().optional().default(true).describe('Show line numbers'),
  head_limit: z.number().int().min(1).optional().default(50).describe('Maximum number of results to return'),
});

export const grepTool: Tool = {
  name: 'grep',
  description: `Fast content search using ripgrep (if available) or fallback.

This is the preferred tool for searching code. Use it instead of running grep via bash when possible.

Features:
- Full regex support
- Filter by glob
- Different output modes (content, files_with_matches, count)
- Automatically respects .gitignore
- Requires ripgrep (rg) on PATH; returns an error if rg is unavailable`,
  parameters: GrepParams,
  safety: {
    sideEffect: 'read',
    permission: 'allow',
    reason: 'Searches file paths and matching lines without modifying files.',
  },
  async execute(args) {
    const params = GrepParams.parse(args);
    const { pattern, path = '.', glob, output_mode, '-i': caseInsensitive, '-n': showLineNumbers, head_limit } = params;

    const cwd = process.cwd();
    const targetPath = path === '.' ? cwd : path;

    // Try ripgrep first (much faster and better)
    try {
      const rgArgs = ['--json', '--no-heading', '--with-filename'];

      if (caseInsensitive) rgArgs.push('-i');
      if (showLineNumbers) rgArgs.push('-n');
      if (glob) rgArgs.push('--glob', glob);

      rgArgs.push('-m', String(head_limit));
      rgArgs.push(pattern, targetPath);

      const { stdout } = await execa('rg', rgArgs, {
        cwd,
        reject: false,
      });

      if (!stdout) {
        return output_mode === 'count' ? '0 matches found' : 'No matches found.';
      }

      // Parse ripgrep JSON output
      const lines = stdout.trim().split('\n');
      const results: any[] = [];

      for (const line of lines) {
        try {
          const parsed = JSON.parse(line);
          if (parsed.type === 'match') {
            results.push(parsed);
          }
        } catch {}
      }

      if (results.length === 0) {
        return 'No matches found.';
      }

      if (output_mode === 'count') {
        return `${results.length} matches found (limited to first ${head_limit})`;
      }

      if (output_mode === 'files_with_matches') {
        const files = new Set(results.map(r => r.data.path.text));
        return Array.from(files).slice(0, head_limit).join('\n');
      }

      // content mode
      const formatted = results.slice(0, head_limit).map((r: any) => {
        const file = r.data.path.text;
        const lineNum = r.data.line_number;
        const text = r.data.lines.text.trim();
        return `${file}:${lineNum}: ${text}`;
      });

      return formatted.join('\n');
    } catch (err) {
      return `Error: ripgrep (rg) is required for grep but was not found on PATH.`;
    }
  },
};

File: internal/tools/grep.go
package tools

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type grepTool struct {
	baseTool
	workspaceRoot string
}

type grepMatch struct {
	file string
	line int
	text string
	hits int
}

func NewGrepTool(workspaceRoot string) Tool {
	return grepTool{
		baseTool: baseTool{
			name:        "grep",
			description: "Search file contents with a regular expression inside the workspace.",
			parameters: Schema{
				Type: "object",
				Properties: map[string]PropertySchema{
					"pattern":          {Type: "string", Description: "Regular expression pattern to search for."},
					"path":             {Type: "string", Description: "Directory or file to search. Defaults to workspace root.", Default: "."},
					"glob":             {Type: "string", Description: `Optional glob filter, for example "**/*.go".`},
					"output_mode":      {Type: "string", Description: "Output mode.", Enum: []string{"content", "files_with_matches", "count"}, Default: "content"},
					"-i":               {Type: "boolean", Description: "Case insensitive search.", Default: false},
					"case_insensitive": {Type: "boolean", Description: "Case insensitive search.", Default: false},
					"head_limit":       {Type: "integer", Description: "Maximum content results to return.", Default: 50, Minimum: intPtr(1)},
				},
				Required:             []string{"pattern"},
				AdditionalProperties: false,
			},
			safety: readOnlySafety("Searches file paths and matching lines without modifying files."),
		},
		workspaceRoot: normalizeWorkspaceRoot(workspaceRoot),
	}
}

func (tool grepTool) Run(_ context.Context, args map[string]any) Result {
	pattern, err := aliasedStringArg(args, []string{"pattern", "query", "regex", "search", "expression"}, "", true, false)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	// Optional with a "." default: treat an explicit empty path (a common
	// weak-model quirk) the same as the key being absent rather than erroring.
	targetPath, err := aliasedStringArg(args, []string{"path", "dir", "directory"}, ".", false, true)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	if targetPath == "" {
		targetPath = "."
	}
	globPattern, err := stringArg(args, "glob", "", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	outputMode, err := stringArg(args, "output_mode", "content", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	if outputMode != "content" && outputMode != "files_with_matches" && outputMode != "count" {
		return errorResult("Error: Invalid arguments for grep: output_mode must be content, files_with_matches, or count")
	}
	caseInsensitive, err := boolArg(args, "case_insensitive", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	shortInsensitive, err := boolArg(args, "-i", false)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}
	headLimit, err := intArg(args, "head_limit", 50, 1, 0)
	if err != nil {
		return errorResult("Error: Invalid arguments for grep: " + err.Error())
	}

	if caseInsensitive || shortInsensitive {
		pattern = "(?i)" + pattern
	}
	compiled, err := regexp.Compile(pattern)
	if err != nil {
		return errorResult("Error running grep: " + err.Error())
	}

	target, _, err := resolveWorkspacePath(tool.workspaceRoot, targetPath)
	if err != nil {
		return errorResult("Error running grep: " + err.Error())
	}

	// Resolve the workspace root through symlinks ONCE so (a) confinement checks
	// and (b) Rel computations both use the canonical root. tool.workspaceRoot is
	// only Abs-normalized (no EvalSymlinks); using it directly would produce
	// "../"-laden relative paths when the root itself lives under a symlink (e.g.
	// macOS /tmp -> /private/tmp) and would not catch files that resolve outside.
	resolvedRoot, err := filepath.EvalSymlinks(tool.workspaceRoot)
	if err != nil {
		return errorResult("Error running grep: " + err.Error())
	}

	var globMatcher *regexp.Regexp
	if globPattern != "" {
		globMatcher, err = compileGlob(globPattern)
		if err != nil {
			return errorResult("Error running grep: " + err.Error())
		}
	}

	files, err := grepFiles(resolvedRoot, target, globMatcher)
	if err != nil {
		return errorResult("Error running grep: " + err.Error())
	}

	matches := collectGrepMatches(resolvedRoot, files, compiled)
	if len(matches) == 0 {
		if outputMode == "count" {
			return okResult("0 matches found")
		}
		return okResult("No matches found.")
	}

	switch outputMode {
	case "count":
		total := 0
		for _, match := range matches {
			total += match.hits
		}
		return okResult(fmt.Sprintf("%d matches found", total))
	case "files_with_matches":
		seen := map[string]bool{}
		files := []string{}
		for _, match := range matches {
			if !seen[match.file] {
				seen[match.file] = true
				files = append(files, match.file)
			}
		}
		sort.Strings(files)
		return okResult(strings.Join(files, "\n"))
	default:
		lines := make([]string, 0, len(matches))
		for _, match := range matches {
			if len(lines) >= headLimit {
				break
			}
			lines = append(lines, fmt.Sprintf("%s:%d: %s", match.file, match.line, match.text))
		}
		return Result{
			Status:    StatusOK,
			Output:    strings.Join(lines, "\n"),
			Truncated: len(matches) > headLimit,
		}
	}
}

// confineGrepFile resolves a candidate file through symlinks and returns its
// clean, slash-separated path RELATIVE to the (already symlink-resolved) root.
// It returns ok=false when the resolved file escapes the workspace root, so a
// symlink inside the workspace that points outside is never searched/returned —
// mirroring resolveWorkspaceTargetPath / read_file confinement. resolvedRoot must
// already be EvalSymlinks-resolved so the Rel result is "../"-free for in-root
// files even when the root lives under a symlink.
func confineGrepFile(resolvedRoot string, path string) (string, string, bool) {
	resolved, err := filepath.EvalSymlinks(path)
	if err != nil {
		return "", "", false
	}
	relative, err := filepath.Rel(resolvedRoot, resolved)
	if err != nil {
		return "", "", false
	}
	if relative == ".." || strings.HasPrefix(relative, ".."+string(filepath.Separator)) || filepath.IsAbs(relative) {
		return "", "", false
	}
	// Return the symlink-resolved absolute path too: callers must read THAT (not
	// the unresolved input) so a symlink swap between this check and the read
	// cannot escape the workspace boundary.
	return filepath.ToSlash(relative), resolved, true
}

func grepFiles(resolvedRoot string, target string, globMatcher *regexp.Regexp) ([]string, error) {
	info, err := os.Stat(target)
	if err != nil {
		return nil, err
	}

	if !info.IsDir() {
		relative, _, ok := confineGrepFile(resolvedRoot, target)
		if !ok {
			return []string{}, nil
		}
		if globMatcher == nil || globMatcher.MatchString(relative) {
			return []string{target}, nil
		}
		return []string{}, nil
	}

	files := []string{}
	err = filepath.WalkDir(target, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if path == target {
			return nil
		}
		if entry.IsDir() && shouldSkipDirectory(entry.Name()) {
			return filepath.SkipDir
		}
		if entry.IsDir() {
			return nil
		}
		// Confine each candidate through symlinks: a symlink inside the workspace
		// pointing to a file OUTSIDE the root must be skipped, not searched.
		relative, _, ok := confineGrepFile(resolvedRoot, path)
		if !ok {
			return nil
		}
		if globMatcher == nil || globMatcher.MatchString(relative) {
			files = append(files, path)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(files)
	return files, nil
}

func collectGrepMatches(resolvedRoot string, files []string, compiled *regexp.Regexp) []grepMatch {
	matches := []grepMatch{}
	for _, file := range files {
		// Re-confine at read time (defense-in-depth) AND to compute the clean
		// workspace-relative path used in output.
		relative, resolvedPath, ok := confineGrepFile(resolvedRoot, file)
		if !ok {
			continue
		}
		// Read the symlink-RESOLVED path that confineGrepFile validated, not the
		// raw candidate, so a symlink swapped in after the check can't escape.
		content, err := os.ReadFile(resolvedPath)
		if err != nil {
			continue
		}
		for index, line := range strings.Split(strings.ReplaceAll(string(content), "\r\n", "\n"), "\n") {
			lineMatches := compiled.FindAllStringIndex(line, -1)
			if len(lineMatches) == 0 {
				continue
			}
			matches = append(matches, grepMatch{
				file: relative,
				line: index + 1,
				text: strings.TrimRight(line, "\r"),
				hits: len(lineMatches),
			})
		}
	}
	return matches
}
